indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java

```diff
@@ -0,0 +1,685 @@
+  {
+    final ServiceMetricEvent.Builder metricBuilder = ServiceMetricEvent.builder();
+    metricBuilder.setDimension("taskActionType", SegmentAllocateAction.TYPE);
+    metricBuilder.setDimension(DruidMetrics.DATASOURCE, key.dataSource);
+    metricBuilder.setDimension(DruidMetrics.INTERVAL, key.preferredAllocationInterval.toString());
+    emitter.emit(metricBuilder.build(metric, value));
+  }
+
+  public void becomeLeader()
+  {
+    if (isEnabled()) {
+      // Start polling the queue
+      log.info("Elected leader. Starting queue processing.");
+      scheduleQueuePoll(maxWaitTimeMillis);
+    } else {
+      log.info("Elected leader but batched segment allocation is disabled.");
+    }
+  }
+
+  public void stopBeingLeader()
+  {
+    log.info("Not leader anymore. Stopping queue processing.");
+  }
+
+  /**
+   * A batch of segment allocation requests.
+   */
+  private class AllocateRequestBatch
+  {
+    private long queueTimeMillis;
+    private final AllocateRequestKey key;
+
+    /**
+     * This must be accessed through methods synchronized on this batch.
+     * It is to avoid races between a new request being added just when the batch
+     * is being processed.
+     */
+    private final Map<SegmentAllocateRequest, CompletableFuture<SegmentIdWithShardSpec>>
+        requestToFuture = new HashMap<>();
+
+    AllocateRequestBatch(AllocateRequestKey key)
+    {
+      log.info("Creating a new batch with key: %s", key);
+      this.key = key;
+    }
+
+    synchronized Future<SegmentIdWithShardSpec> add(SegmentAllocateRequest request)
+    {
+      log.info("Adding request to batch [%s]: %s", key, request.getAction());
+      return requestToFuture.computeIfAbsent(request, req -> new CompletableFuture<>());
+    }
+
+    synchronized void merge(AllocateRequestBatch batch)
+    {
+      requestToFuture.putAll(batch.requestToFuture);
+      batch.requestToFuture.clear();
+    }
+
+    synchronized List<SegmentAllocateRequest> getRequests()
+    {
+      return new ArrayList<>(requestToFuture.keySet());
+    }
+
+    synchronized void markCompleted()
+    {
+      if (!requestToFuture.isEmpty()) {
+        log.info("Marking [%d] requests in batch [%s] as failed.", size(), key);
+        requestToFuture.values().forEach(future -> future.complete(null));
+        requestToFuture.keySet().forEach(
+            request -> emitTaskMetric("task/action/failed/count", 1L, request)
+        );
+        requestToFuture.clear();
+      }
+    }
+
+    synchronized void handleResult(SegmentAllocateResult result, SegmentAllocateRequest request)
+    {
+      request.incrementAttempts();
+
+      if (result.isSuccess()) {
+        emitTaskMetric("task/action/success/count", 1L, request);
+        requestToFuture.remove(request).complete(result.getSegmentId());
+        return;
+      }
+
+      log.info("Failed to allocate segment for action [%s]: %s", request.getAction(), result.getErrorMessage());
```

Comment:
```suggestion
      log.warn("Failed to allocate segment for action [%s]: %s", request.getAction(), result.getErrorMessage());
```

Reply:
can this action be anything else other than SegmentAllocationAction?

Reply:
No, this queue is only for batching `SegmentAllocateAction`s only.

Reply:
I think I will just remove this log line and include the error message in the subsequent logs inside the if-else.

Reply:
Updated.